### ISSUE TITLE
Random crashes in some games (Vulkan)

### ISSUE THREAD

**Issue by leopard20** [hrydgard/ppsspp#10811]
What happens?
Faltout crashes sometimes in the middle of a race.
This is the debug error my system gave me:

```
backtrace:
    #00 pc 0000000000630710  /data/app/org.ppsspp.ppsspp-1/lib/arm64/libppsspp_jni.so (_ZN17VulkanQueueRunner13GetRenderPassE19VKRRenderPassActionS0_S0_13VkImageLayoutS1_S1_+204)
    #01 pc 000000000062a88c  /data/app/org.ppsspp.ppsspp-1/lib/arm64/libppsspp_jni.so (_ZN4Draw9VKContext15GetNativeObjectENS_12NativeObjectE+84)
    #02 pc 0000000000499a04  /data/app/org.ppsspp.ppsspp-1/lib/arm64/libppsspp_jni.so (_ZN16DrawEngineVulkan7DoFlushEv+4060)
    #03 pc 00000000004f4d9c  /data/app/org.ppsspp.ppsspp-1/lib/arm64/libppsspp_jni.so (_ZN9GPUCommon18Execute_ViewMtxNumEjj+200)
    #04 pc 00000000004f7b8c  /data/app/org.ppsspp.ppsspp-1/lib/arm64/libppsspp_jni.so (_ZN9GPUCommon11FastRunLoopER11DisplayList+252)
    #05 pc 00000000004f7624  /data/app/org.ppsspp.ppsspp-1/lib/arm64/libppsspp_jni.so (_ZN9GPUCommon13InterpretListER11DisplayList+848)
    #06 pc 00000000004f6a10  /data/app/org.ppsspp.ppsspp-1/lib/arm64/libppsspp_jni.so (_ZN9GPUCommon14ProcessDLQueueEv+112)
    #07 pc 00000000004f6920  /data/app/org.ppsspp.ppsspp-1/lib/arm64/libppsspp_jni.so (_ZN9GPUCommon11EnqueueListEjji10PSPPointerI13PspGeListArgsEb+1140)
    #08 pc 000000000038a618  /data/app/org.ppsspp.ppsspp-1/lib/arm64/libppsspp_jni.so (_Z10WrapU_UUIUIXadL_Z16sceGeListEnQueuejjijEEEvv+60)
    #09 pc 0000000000368e04  /data/app/org.ppsspp.ppsspp-1/lib/arm64/libppsspp_jni.so (_Z23CallSyscallWithoutFlagsPK11HLEFunction+32)
    #10 pc 000000000000e95c  <anonymous:0000007f54800000>
```
Any idea what it means?!

What hardware and operating system are you running PPSSPP on? GPU might matter if it's a graphical issue.
Android 7.0 
Vulkan
Snapdragon 821
v1.5.4-773

**Comment by weihuoya** [comment]
GoW met this issue sometime.

this cause by hashmap, when get renderpass, hashmap return null, but when set renderpass, hashmap assert there was one.

is this cause by hashmap grow function?

**Comment by hrydgard** [comment]
Ugh, nasty. Gonna have to look at this.

**Comment by leopard20** [comment]
@hrydgard 
This is not limited to Flatout. I got similar crashes in other games too.
This one is from The 3rd Birthday:
```
*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
Build fingerprint: 'Xiaomi/natrium/natrium:7.0/NRD90M/8.1.11:user/release-keys'
Revision: '0'
ABI: 'arm64'
pid: 2652, tid: 2676, name: AndroidRender  >>> org.ppsspp.ppsspp <<<
signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0xd
    x0   0000000000000044  x1   0000007f58d1fb28  x2   0000000000000010  x3   00000000000000bf
    x4   7b02000000000000  x5   00000000779b1000  x6   000000c400000000  x7   00000000c4000000
    x8   0000000000000000  x9   00000000ec703be9  x10  00000000165667b1  x11  0000000000000018
    x12  0000000000000001  x13  00000000ec71e30a  x14  0000007f5cc83108  x15  00000000fcc6cf41
    x16  0000007f5fc30a98  x17  0000007f7cb9cdc0  x18  0000000024282081  x19  0000007f58d1fa00
    x20  0000007f5f89c469  x21  0000007f5f74f6b5  x22  000000000000001f  x23  000000000000000d
    x24  0000000000000018  x25  0000007f7b468bd8  x26  000000000000000b  x27  0000007f5f89c42f
    x28  0000007f5f89c44f  x29  0000007f5cc831e0  x30  0000007f5f2667c8
    sp   0000007f5cc83020  pc   0000007f5f266798  pstate 0000000020000000

backtrace:
    #00 pc 0000000000608798  /data/app/org.ppsspp.ppsspp-2/lib/arm64/libppsspp_jni.so (_ZN17VulkanQueueRunner13GetRenderPassE19VKRRenderPassActionS0_S0_13VkImageLayoutS1_S1_+204)
    #01 pc 0000000000603314  /data/app/org.ppsspp.ppsspp-2/lib/arm64/libppsspp_jni.so (_ZN4Draw9VKContext15GetNativeObjectENS_12NativeObjectE+84)
    #02 pc 000000000048a78c  /data/app/org.ppsspp.ppsspp-2/lib/arm64/libppsspp_jni.so (_ZN16DrawEngineVulkan7DoFlushEv+1788)
    #03 pc 00000000004e5ad8  /data/app/org.ppsspp.ppsspp-2/lib/arm64/libppsspp_jni.so (_ZN9GPUCommon13InterpretListER11DisplayList+620)
    #04 pc 00000000004e4fa8  /data/app/org.ppsspp.ppsspp-2/lib/arm64/libppsspp_jni.so (_ZN9GPUCommon14ProcessDLQueueEv+112)
    #05 pc 00000000004e54b4  /data/app/org.ppsspp.ppsspp-2/lib/arm64/libppsspp_jni.so (_ZN9GPUCommon11UpdateStallEij+84)
    #06 pc 0000000000382890  /data/app/org.ppsspp.ppsspp-2/lib/arm64/libppsspp_jni.so
    #07 pc 0000000000360180  /data/app/org.ppsspp.ppsspp-2/lib/arm64/libppsspp_jni.so (_Z23CallSyscallWithoutFlagsPK11HLEFunction+32)
    #08 pc 0000000000056a80  <anonymous:0000007f55800000>
```

**Comment by hrydgard** [comment]
Interesting. I might add some debug logging later, I assume you can get adb logs.

**Comment by leopard20** [comment]
Yes. But I've never done it before. Does provide more info than above?

**Comment by hrydgard** [comment]
I think I know what's going on here, the render pass cache can get accessed from two different threads and if it's being grown at the same time that it's being accessed, it'll blow up. I'll fix it.

**Comment by hrydgard** [comment]
Hopefully that will take care of it.

**Comment by leopard20** [comment]
Will test it soon.
This is the logcat I dumped directly from my terminal:
[logcat.txt](https://github.com/hrydgard/ppsspp/files/1852078/logcat.txt)

Not sure if it made any difference if I did it via adb.

**Comment by hrydgard** [comment]
Thanks! Unfortunately the app logs had been shifted out of view, not that I expected to see much.

But looking back a bit in your logs I see a different message, which is also interesting:

``` 03-26 22:38:53.710 30135 30135 F DEBUG   : Abort message: 'VulkanMemory.cpp:176 (Destroy): [slabUsage == 2] Destroy: slabUsage has unexpected value 1'  ```

and 

```[!slabs_.empty()] No slabs - can't be anything to free! double-freed?'```

Those are unrelated to this crash though, seems to be a UI image issue related to the savestate screenshots views in the pause screen. very interesting...

**Comment by leopard20** [comment]
@hrydgard 
Yeah. Some of my save states just show a black screen. Is the error related to this?

**Comment by hrydgard** [comment]
Is that specific to certain games? Wouldn't expect that though...

**Comment by weihuoya** [comment]
the savestate screenshots may delete in Gallery, and it is black in pause screen.
should add a .nomedia file in /sdcard/PSP

**Comment by leopard20** [comment]
@hrydgard 
Yes it only happens in a few games. 

@weihuoya 
I don't delete them from Gallery! I'm not a noob! 😉 But yeah a .nomedia file should be useful.

**Comment by leopard20** [comment]
@hrydgard 
I noticed that some of my saved states didn't even have a screenshot! But I can swear I never deleted the images! They probably were never created! (3 of which are from the 3rd Birthday)

I noticed that Patapon has images but they appear black in the menu:
![screenshot_2018-03-27-18-54-35-281_org ppsspp ppsspp](https://user-images.githubusercontent.com/16115820/37973690-54dd8b4a-31f0-11e8-9c18-ee8c58db0f12.png)

Here are the screenshots:
![ucus98711_1_1521822959582 00_1](https://user-images.githubusercontent.com/16115820/37973776-83688d8e-31f0-11e8-9c3c-4dd3bd40ea98.jpg)
![ucus98711_1_1521822959595 00_0](https://user-images.githubusercontent.com/16115820/37973790-8a59b9c4-31f0-11e8-887e-454b88b92ece.jpg)

This one is black and from Assassin's Creed:
![ulus10391_1 01_0](https://user-images.githubusercontent.com/16115820/37973843-ac37a9de-31f0-11e8-9d0e-08ff08eeacb3.jpg)

They all appear blank (not black!) in savedata manager.
